File: 121.java
class Solution {
    public int maxProfit(int[] prices) {
        if (prices.length == 0) return 0;
        int min = prices[0];
        int res = 0;
        for (int i = 1; i < prices.length; i++) {
            min = Math.min(min, prices[i]);
            res = Math.max(res, prices[i]-min);
        }
        return res;
    }
}

File: 121.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) <= 1: return 0
        diff = [0] * (len(prices)-1)
        for i in range(len(diff)):
            diff[i] = prices[i+1] - prices[i]
        cusum = [diff[0]] * len(diff)
        for i in range(1,len(diff)):
            cusum[i] = max(cusum[i-1], 0) + diff[i]
        return max(max(cusum),0)